share/wake/lib/system/http.wake

```diff
@@ -0,0 +1,199 @@
+    # content-length: 54
+    # other: header
+    # other: header
+    #
+    # Expected request with `Content-Type: application/json`
+    require Pass (CurlResponseParts codeline headers body) =
+        def parts = tokenize `\n` str 
+
+        require code, rest = parts
+        else failWithError "Unexpected empty curl response"
+
+        # Check if only the header line was provided
+        require False = rest.len == 0
+        else (Pass (CurlResponseParts code Nil ""))
+
+        # each header is on its own line, headers stop when we reach an empty line
+        def loop = match _
+            Nil -> Nil
+            "", _ -> Nil
+            x, xs -> x, loop xs
+
+        def headers = loop rest
+
+        # The body is just the remaining lines minus the header lines and empty line
+        def body = 
+            rest
+            | drop (headers.len + 1)
+            | catWith "\n"
+
+        Pass (CurlResponseParts code headers body)
+
+
+    require Pass status =
+        require (code, Nil) = extract `^HTTP\/[\d\.]+ (\d+) .*$` codeline 
+        else failWithError "unable to extract status code from '{codeline}'"
+
+        require (Some x) = code | (intbase 10)  
+        else failWithError "unable to parse {code} as a base 10 integer"
+
+        Pass x
+
+    require Pass headers = 
+        def loop header = 
+            require (name, value, Nil) = extract `^([-_A-Za-z0-9]+):\s*(.+)$` header 
+            else failWithError "Unable to extract header from '{header}'"
+
+            Pass (HttpHeader name value)
+
+        headers
+        | map loop
+        | findFail
+
+
+    Pass (HttpResponse status headers body)
+
+
+# Makes the request specified by request and parses into a HttpResponse
+#
+# WARNING: This function will likely break the sandbox and make your build unreliable
+export def makeRequest (request: HttpRequest): Result HttpResponse Error =
+    def methodToString = match _
+        HttpMethodGet -> "GET"
+        HttpMethodHead -> "HEAD"
+        HttpMethodPost -> "POST"
+        HttpMethodPut -> "PUT"
+        HttpMethodDelete -> "DELETE"
+        HttpMethodConnect -> "CONNECT"
+        HttpMethodOptions -> "OPTIONS"
+        HttpMethodTrace -> "TRACE"
+        HttpMethodPatch -> "PATCH"
+
+    def headerToCurlFlag (HttpHeader name value) = 
+        "--header", "{name}:{value}", Nil
+
+    def headers =
+        request.getHttpRequestHeaders
+        | map headerToCurlFlag
+        | flatten
+
+    def method = request.getHttpRequestMethod.methodToString
+
+    # wake-format off
+    def cmd =
+        "curl",
+        "-i",
+        "--request", method,
+        "--data", request.getHttpRequestBody,
+        "--url", request.getHttpRequestUrl,
+        headers
+
+    require Pass stdout =
+        makeExecPlan cmd Nil
+        | setPlanLabel "http: {method} {request.getHttpRequestUrl}"
+        # Curl dumps download stats to stderr
+        | setPlanStderr logNever
```

Comment:
Might still be helpful sometimes?  Or does our virtual TTY throw curl's formatting off?
```suggestion
        | setPlanStderr logInfo
```

Reply:
I don't feel strongly about this, main argument against imo is that we should be hitting this _a lot_ and logging this info doesn't see valuable enough for the extra time it takes to write it to the stream since it'll be in the database anyways

Reply:
`logInfo` should only be visible if people run with `-v`, right?  Does the performance get hit even if it's not shown?  But, yeah, I was forgetting that people should probably just be looking it up in the database.